### ISSUE TITLE
LightTheme and DarkTheme are being deprecated without any replacement

### ISSUE THREAD

**Issue by williambrode** [fyne-io/fyne#4208]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

From the code:

// LightTheme defines the built-in light theme colors and sizes.
//
// Deprecated: This method ignores user preference and should not be used, it will be removed in v3.0.

However, I can't see any obvious replacement which would allow me to change the theme variant for the user.  I have a settings menu where they can select Light or Dark, just as the fyne demo has.  Fyne demo even still uses the deprecated function.  I believe the comment should be updated to suggest another way to set the theme variant, so users can keep the functionality once its removed. 

### How to reproduce

Use LightTheme(), once its removed, dunno how to change theme variant.

### Example code

From fyne demo

```go
	themes := container.NewGridWithColumns(2,
		widget.NewButton("Dark", func() {
			a.Settings().SetTheme(theme.DarkTheme())
		}),
		widget.NewButton("Light", func() {
			a.Settings().SetTheme(theme.LightTheme())
		}),
	)
```

### Fyne version

latest

### Go compiler version

1.19.1

### Operating system and version

Windows 10

**Comment by Jacalz** [comment]
I believe that the answer you seek can be found here: https://github.com/fyne-io/fyne/discussions/4188

**Comment by williambrode** [comment]
That answer does work for me (though probably not everyone).  However, the reason I opened it as an issue is so we can put that information into the comment in the code.  Can we put the link in your post into the //Deprecated comment in code?

**Comment by andydotxyz** [comment]
I'm not sure what sort of hint we should put there. It was removed due to generally being a bad user experience.
I suppose we can hint towards custom themes?

**Comment by williambrode** [comment]
In my case I just wanted to set the theme variant to dark, when the user requested that the theme be dark.  I feel like this should be allowed?  I'm not sure if the way it had been implemented perhaps overwrote other user setting (it looked perhaps like it), but is there no workaround for doing what I want?

**Comment by andydotxyz** [comment]
Hopefully the docs proposed in #4215 avoid this confusion/frustration?

**Comment by williambrode** [comment]
What I'm still not understanding is why the assumption is that developers are trying to override a user's theme preference.  I'm only trying to help a user to set their theme variant.

What is the suggested way for a user to set the their theme variant if not through the app itself?  From reading the code in https://github.com/fyne-io/fyne/blob/master/cmd/fyne_settings/settings/appearance.go it looks like we are supposed to load a well-known settings file, edit the settings, then write it back to the file?

Perhaps the easiest thing (for developers) would be to to add a SetUserThemeVariant() func to that module which would be a direct replacement of .SetTheme()?  So it would load the file, change the variant and write it back out for the user.  I know that is more code to write - but I was just trying to explain what I think most developers might be thinking when using the existing SetTheme() method.

Thanks for all you do!

**Comment by andydotxyz** [comment]
We assume that the user will have their preference set at the OS level - that is what the apps aim to match.

Surely users are looking for consistency not a per-app theme variant setting. Most software that has theme choice in it added that before OS finally standardised on supporting light/dark mode.

**Comment by andydotxyz** [comment]
P.s. no users who do want it per app don't edit files - you are welcome to provide whatever theme overrides you want in your app. You can see in the PR I linked that it's not difficult and it certainly doesn't involve writing global settings files manually.

**Comment by williambrode** [comment]
The reason I added this to my app was because a user specifically requested that they be able to change my app to "Light" theme even though their OS default was dark.  So I think that is a reasonable use case.

So the point here was not that there was a better way to set the theme for the user, but rather that you would prefer to make it harder to do?  I'm not convinced that path is best.  Yes, I'm sure I can add the same logic you added in fyne_demo - but it feels a bit like a regression.

**Comment by andydotxyz** [comment]
Are you able to enquire of the user why they want this app to be different to their OS colouring? It is very helpful to know why a change is requested in case there is a better solution/naming.

The logic behind the change is that we have one builtin theme - and it matches user preference. If your app wants to control theming then it should use a custom theme. As you can see it's not difficult, but app developers should own the fact that they are asserting an override of system settings. Just like the text size - we match the system, to make it configurable in the main package does not make sense but a custom theme can override it.

**Comment by fuskiid** [comment]
> Are you able to enquire of the user why they want this app to be different to their OS colouring? It is very helpful to know why a change is requested in case there is a better solution/naming.
> 
> The logic behind the change is that we have one builtin theme - and it matches user preference. If your app wants to control theming then it should use a custom theme. As you can see it's not difficult, but app developers should own the fact that they are asserting an override of system settings. Just like the text size - we match the system, to make it configurable in the main package does not make sense but a custom theme can override it.

In terms of user preference/precedent, almost all apps that I see used (in current day) that allow theming, also allow the user to override OS light/dark preference in-app; I believe this is also how the current fyne-demo operates.

Generally the setting will start on "automatic" which chooses default theme based upon OS, but will allow user to manually override light/dark mode instead. 

Discord: {light, dark, sync-with-computer}
Matrix (Element): {light, dark, "Match system theme"}
Signal (Messenger): {light, dark, "System"}
etc...

I don't understand why anyone would prefer light-mode on anything, but I have friends that do ( crazy people ;) ) and the expectation is there that there should be a per-app override based upon the apps purpose and user preference for that individual purpose.

I (as the developer) don't want to override the users OS preference, I just want to allow the user to do so themself if they so desire.

Regardless, due to fyne's custom theming ability I am sure I can make my own "light, dark" themes and do this process myself, I just find having them ready-to-go with library easier to use.

**Comment by andydotxyz** [comment]
> In terms of user preference/precedent, almost all apps that I see used (in current day) that allow theming, also allow the user to override OS light/dark preference in-app; I believe this is also how the current fyne-demo operates.

Agreed, thanks. If you want custom theming then you can add light/dark variant option (or still leave it to the system preference...). But simply choosing to override OS without any actual theme capabilities isn't so useful, so we discourage it.

**Comment by dweymouth** [comment]
I think this can be closed, as the replacement is loading an app-defined custom theme

**Comment by bob-anderson-ok** [comment]
I don't know how to apply the suggested code replacement. Please help.

(I think you should not deprecate those functions. Fyne has definite "opinions" - this need not be one of them.)
`
// Simple user selectable theme change currently in place, but with deprecated functions
func changeTheme(checked bool) {
	if checked {
		myWin.App.Settings().SetTheme(theme.LightTheme())
	} else {
		myWin.App.Settings().SetTheme(theme.DarkTheme())
	}
}

// Suggested code to replace deprecated functions

func DarkTheme(fallback fyne.Theme) fyne.Theme {
	return &forcedVariant{Theme: fallback, variant: 0} // avoid import loops
}

func LightTheme(fallback fyne.Theme) fyne.Theme {
	return &forcedVariant{Theme: fallback, variant: 1} // avoid import loops
}

type forcedVariant struct {
	fyne.Theme

	variant fyne.ThemeVariant
}

func (f *forcedVariant) Color(name fyne.ThemeColorName, _ fyne.ThemeVariant) color.Color {
	return f.Theme.Color(name, f.variant)
}

// But I don't see how to apply the new code


`

**Comment by andydotxyz** [comment]
You can see it in use at https://github.com/fyne-io/fyne/blob/ffb34ca74c1029e32ee543d0b603d87189acaeb7/cmd/fyne_demo/main.go#L249

Please don't comment on closed issues - we have many support channels open to ask questions that will be answered more quickly.

**Comment by andydotxyz** [comment]
> (I think you should not deprecate those functions. Fyne has definite "opinions" - this need not be one of them.)

> (I think you should not deprecate those functions. Fyne has definite "opinions" - this need not be one of them.)

Our API aims for best practice, and allowing devs to override user choice for theme variant is not best practice. Users can select the theme through their OS, the fyne_settings app or using the FYNE_THEME environment - coding it into your app will ignore all of these features.

**Comment by bob-anderson-ok** [comment]
Andy:

Those choices are all more difficult for the user than simply clicking a
checkbox.

You could add a "theme selector" widget which defaulted automatically to
the users "global' preference but allows an alternative to be chosen
for the particular app. If an alternate were selected, the programmer
should persist it through Preferences. Or the widget could automatically
persist it if the window had an ID.

I'm not proposing that devs override user choice. On the contrary, it is to
provide a way for the user to make fine-grained, easier choices.

Thanks for your response. This is my first use of Golang and Fyne. Except
for some frustrations with 'type tangle', I'm having a good experience.

Look for me to become a Fyne supporter. I want this project to succeed long
term. I hope that you can find a way to monetize it in a reasonable way to
ensure its longevity.

Bob Anderson

**Comment by andydotxyz** [comment]
> You could add a "theme selector" widget which defaulted automatically to
> the users "global' preference but allows an alternative to be chosen
> for the particular app. If an alternate were selected, the programmer
> should persist it through Preferences. Or the widget could automatically
> persist it if the window had an ID.

Yes, some people have done this. If you are offering theme switching then the new code is far more suitable - as you can switch between any number of custom themes. The previous was just light or dark modes of builtin.

We have not removed any functionality, just made it harder to do what was (commonly) not best practice usage.

**Comment by atljoseph** [comment]
I’m a bit lost. It was removed because a bad user experience. Aren’t developers users in this sense? Can’t a bad user experience also defined as deprecating something and giving absolutely no example about how to best employ the removed feature in a the preferred way? I’m trying to retain my sanity here and bite my tongue. Wouldn’t best practice be helping folks not flail around and feel lost? Can you address THAT user experience? Every lib I know of spells out in the deprecation comment how to achieve the thing being deprecated in a different way. Are you wishing confusion on folks? I know you aren’t but that is the net result.

**Comment by andydotxyz** [comment]
When we talk about users and UX we are referring to the end users of apps created with Fyne. Users of the API are referred to developers and their consideration is called developer experience (DX).

These APIs were removed as documented because developers were creating apps that ignored user preference because they would code the developer preference in (possibly by mistake?).

All apps are welcome to provide a custom theme and a theme is able to express the prohibition of either light or dark variants to force the developer or brand colours. There are many examples of this but it is not a direct replacement. There is no direct replacement as user preference should never have been so simply overridden in this way.

**Comment by andydotxyz** [comment]
> We assume that the user will have their preference set at the OS level - that is what the apps aim to match.

For clarity this was the reason for deprecation and the approach we continue to encourage.

**Comment by Jacalz** [comment]
Maybe https://github.com/fyne-io/fyne/blob/develop/internal/test/theme.go (or the equivalent code in fyne_demo) should be linked to in the deprecation message?

**Comment by atljoseph** [comment]
Deprecation messages are also one of the best ways to steer people towards proper usage when things are in flux. A deprecation message without a clear explanation of the preferred alternative is not helpful. It may be clear to the person writing the deprecation message, but less clear to others. It represents yet one more thing to figure out for folks who have committed to using fyne. Meanwhile, the answer is known, and thought about, and debated… Why not share that knowledge? Why not share that feeling of certainty? In exchange, the thing you have built will be that much easier to understand. As a result, it will be easier to use. And apps that employ it will become more prominent.

**Comment by andydotxyz** [comment]
Thanks for your thoughts. Of course we are always looking to make developer experience better - it is a founding principle of the project.

If you have suggestions about what the message should say to that end please feel free to do so. It could even be a PR to include the message, as everything here is open source and hosted on GitHub.
